core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1121,8 +1121,16 @@ class ReplicaManager(val config: KafkaConfig,
           hwThreadInitialized = true
         }
 
+        // from existing partitions for which we have become the leader, select those with a future replica
+        val invalidatedPartitionsWithFutureReplicas = partitionsBecomeLeader.filter(_.futureLocalReplica.isDefined)
+
+        // remove selected partitions from alter log dirs fetching; they will be re-added below
+        // this avoids a scenario where the local replica becomes fenced when fetching from the future replica as leader
+        replicaAlterLogDirsManager.removeFetcherForPartitions(
+          invalidatedPartitionsWithFutureReplicas.map(_.topicPartition).toSet)
+
         val futureReplicasAndInitialOffset = new mutable.HashMap[TopicPartition, InitialFetchState]
-        for (partition <- newPartitions) {
+        for (partition <- newPartitions ++ invalidatedPartitionsWithFutureReplicas) {
           val topicPartition = partition.topicPartition
           if (logManager.getLog(topicPartition, isFuture = true).isDefined) {
             partition.localReplica.foreach { replica =>
```

Comment:
The code below sets `initOffset` in `InitialFetchState` to high watermark of local replica, which works in case where partition is new. For the case where partition is not new, and we are removing/adding fetcher thread, we need to use future replica's high watermark as `initOffset`. In case where OffsetsForLeaderEpoch return unknown epoch and offset (e.g., using old inter-broker protocol version), the fetch offset will be reset to high watermark , and using high watermark of the local partition may result in future replica skipping some records.

Reply:
Actually, we should set `initOffset` to futureReplica.highWatermark.messageOffset in any case, even if partition is new (although for new partition we would not see the difference, just one less `if` statement).